rfcs/0191-worker-manager-launch-configs.md

```diff
@@ -0,0 +1,122 @@
+# RFC 0191 - Worker Manager launch configurations
+* Comments: [#0191](https://github.com/taskcluster/taskcluster-rfcs/pull/191)
+* Proposed by: @lotas
+
+## Summary
+
+This proposal outlines enhancements to the worker-manager service that would address the following:
+
+* make worker pools launch configurations first-class citizens in the worker-manager service
+* balanced distribution of workers across multiple launch configurations (regions)
+* dynamic adjustment of launch configuration likelihood to be provisioned based on **health metrics**
+
+**Health metrics** - is the combination of the number of fail-to-success provisioning attempts ratio, the number of workers currently running.
+
+## Motivation
+
+The current worker-manager lacks the ability to deal with broken configurations, monitor the health of launch configurations and the availability of cloud resources.
+
+This often leads to the creation of workers that are stuck in provisioning or fail to become operational, wasting resources and reducing system efficiency.
+
+By introducing health tracking and dynamic adjustment mechanisms, we aim to optimize the provisioning process and enhance system reliability.
+
+## Details
+
+New entity will be introduced to worker-manager - **Launch Configuration**.
+Each worker pool will have zero or more **launch configurations** associated with it.
+
+Each **launch configuration** will be assigned a unique ID (hash of its properties), stored in the database and will be immutable (except for the status flag).
+
+Generated ID will include the following properties:
+
+* top-level config properties (`minCapacity`, `maxCapacity`, `scalingRatio`, `lifecyle`)
+* launch configuration itself
+
+During updates of the worker pool configurations, all existing launch configurations that are not present in the new configuration will be marked as `archived`. Launch configurations that have same unique ID (hash) will be kept active. All other will be created as new launch configurations and marked as active.
+
+### Launch Configuration weight
+
+Provisioned workers would be associated with a specific launch configuration (`worker.launchConfigurationId`).
+This will allow us to to know how many workers with this launch configuration were successfully registered and claimed work.
+
+Each launch configuration will have a dynamic `weight` property that will be adjusted automatically based on the following events and metrics:
+
+* total number of successful worker provisioning attempts / registrations
```

Comment:
This talks about total numbers of success/failure, but the summary at the top talks about using a ratio. Is this different somehow? (It seems ideal to me to use a ratio here, otherwise the influence this has will differ a lot based on the size of a worker pool.)

Reply:
Good point. 

My initial concern was to slow down the provisioning for brand new configs, that have not yet seen any successful workers. So in that context, having "total X good"  would be a valuable info to know if it's fine to go full speed. 

Maybe I'm mixing too much in here? Weight distribution between configs and initial provisioning rate for new configs. Probably worth separating those and also keep out the initial provisioning rate out of the scope, since it's not yet clear how to handle this properly.